File: comparing length of columns with stack.py
import math
class Tower:
    def __init__(self, a, b, c):
        self.a = a  
        self.b = b  
        self.c = c  
        self.left = 0 
        self.right = 0 
n = int(input()) 
output = []  # To store all the results
for _ in range(n):
    m = int(input()) 
    towers = []
    for i in range(m):
        a, b, c = map(int, input().split())
        towers.append(Tower(a, b, c))
    stack_left = []  # Stack to calculate leftsum
    stack_right = []  # Stack to calculate rightsum
    leftsum = 0
    rightsum = 0
    result = []
    for i in range(m):  # Single loop to calculate both left and right values
        # Update leftsum
        tower = towers[i]
        tower.left = leftsum
        while stack_left and stack_left[-1].a < tower.a:
            leftsum -= stack_left.pop().c
        stack_left.append(tower)
        leftsum += tower.c

        mirrored_index = m - 1 - i
        mirrored_tower = towers[mirrored_index]
        mirrored_tower.right = rightsum
        while stack_right and stack_right[-1].a < mirrored_tower.a:
            rightsum -= stack_right.pop().c
        stack_right.append(mirrored_tower)
        rightsum += mirrored_tower.c
    for i in range (m):    
        result.append(math.ceil(towers[i].b / (towers[i].left + towers[i].right)))

    output.append(" ".join(map(str, result)))  # Add the result of the current test case to the output

# Print all the results in one go for efficiency
print("\n".join(output))

File: comparing length of columns with stack.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <string>
using namespace std;

struct Tower {
    int a, b, c;
    long long left, right;

    Tower(int a, int b, int c) : a(a), b(b), c(c), left(0), right(0) {}
};

// Manual stack implementation
template <typename T>
class Stack {
    vector<T> elems;

public:
    void push(const T& elem) {
        elems.push_back(elem);
    }

    void pop() {
        if (!elems.empty()) {
            elems.pop_back();
        }
    }

    T& top() {
        return elems.back();
    }

    bool empty() const {
        return elems.empty();
    }
};

int main() {
    int n;
    cin >> n;
    vector<string> output;

    for (int t = 0; t < n; ++t) {
        int m;
        cin >> m;
        vector<Tower> towers;
        Stack<Tower*> tower_stack;
        long long leftsum = 0;

        for (int i = 0; i < m; ++i) {
            int a, b, c;
            cin >> a >> b >> c;
            Tower* tower = new Tower(a, b, c);
            towers.push_back(*tower); 
        }
        
        for (int i = 0; i < m; ++i) {
            Tower& tower = towers[i];
            tower.left = leftsum;
            while (!tower_stack.empty() && tower_stack.top()->a < tower.a) {
                leftsum -= tower_stack.top()->c;
                tower_stack.pop();
            }
            tower_stack.push(&tower);
            leftsum += tower.c;
        }

        while (!tower_stack.empty()) tower_stack.pop();

        long long righsum = 0;
        vector<int> result;

        for (int i = m - 1; i >= 0; --i) {
            Tower& tower = towers[i];
            tower.right = righsum;

            while (!tower_stack.empty() && tower_stack.top()->a < tower.a) {
                righsum -= tower_stack.top()->c;
                tower_stack.pop();
            }
            tower_stack.push(&tower);
            righsum += tower.c;

            result.push_back((int)ceil((double)tower.b / (tower.left + tower.right)));
        }

        string res_str = "";
        for (int i = result.size() - 1; i >= 0; --i) {
            res_str += to_string(result[i]);
            if (i != 0) res_str += " ";
        }
        output.push_back(res_str);
    }

    for (const string& res : output) {
        cout << res << endl;
    }

    return 0;
}
